quantum/debounce/sym_pk.c

```diff
@@ -0,0 +1,120 @@
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+Basic symmetric per-key algorithm. Uses an 8-bit counter per key.
+When no state changes have occured for DEBOUNCE milliseconds, we push the state.
+*/
+
+#include "matrix.h"
+#include "timer.h"
+#include "quantum.h"
+#include <stdlib.h>
+
+#ifndef DEBOUNCE
+#    define DEBOUNCE 5
+#endif
+
+#if (MATRIX_COLS <= 8)
+#    define ROW_SHIFTER ((uint8_t)1)
+#elif (MATRIX_COLS <= 16)
+#    define ROW_SHIFTER ((uint16_t)1)
+#elif (MATRIX_COLS <= 32)
+#    define ROW_SHIFTER ((uint32_t)1)
+#endif
+
+#define debounce_counter_t uint8_t
+
+static debounce_counter_t *debounce_counters;
+static bool                counters_need_update;
+
+#define DEBOUNCE_ELAPSED 251
+#define MAX_DEBOUNCE (DEBOUNCE_ELAPSED - 1)
+
+static uint16_t custom_wrap_timer_reference  = 0;
+static uint8_t  custom_wrap_timer_last_value = 0;
+
+static uint8_t custom_wrap_timer_read(void) {
+    uint16_t custom_wrap_timer_new_reference = timer_read();
+    uint16_t diff                            = custom_wrap_timer_new_reference - custom_wrap_timer_reference;
+    custom_wrap_timer_reference              = custom_wrap_timer_new_reference;
+    custom_wrap_timer_last_value             = (custom_wrap_timer_last_value + diff) % (MAX_DEBOUNCE + 1);
+    return custom_wrap_timer_last_value;
+}
+
+void update_debounce_counters(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, uint8_t current_time);
+void transfer_matrix_values(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, uint8_t current_time);
+
+// we use num_rows rather than MATRIX_ROWS to support split keyboards
+void debounce_init(uint8_t num_rows) {
+    debounce_counters = (debounce_counter_t *)malloc(num_rows * MATRIX_COLS * sizeof(debounce_counter_t));
+    int i             = 0;
+    for (uint8_t r = 0; r < num_rows; r++) {
+        for (uint8_t c = 0; c < MATRIX_COLS; c++) {
+            debounce_counters[i++] = DEBOUNCE_ELAPSED;
+        }
+    }
+}
+
+void debounce(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, bool changed) {
+    uint8_t current_time = custom_wrap_timer_read();
+    if (counters_need_update) {
+        update_debounce_counters(raw, cooked, num_rows, current_time);
+    }
+
+    if (changed) {
+        transfer_matrix_values(raw, cooked, num_rows, current_time);
+    }
+}
+
+// If the current time is > debounce counter, set the counter to enable input.
+void update_debounce_counters(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, uint8_t current_time) {
```

Comment:
This function checks if an unstable cell became stable, and if so, pushes its output.
So the comment `// If the current time is > debounce counter, set the counter to enable input.`
should be updated. 
Maybe comment: `//If  a cell is marked as unstable, we check for stable input until DEBOUNCE_ELAPSED. If it hasn't changed during this time, we push its value to cooked`
And a function name like "update_unstable_cells" might be good.

Reply:
Agreed, comment and function name is left from eager_pk, from which I branched to create sym_pk. It made sense for eager_pk, but due to my modifications it doesn't make sense anymore. Will remove the comment(s), and rename the functions, for it to make sense, commit coming in a few minutes.